contracts/zero-ex/contracts/src/features/libs/LibSignedCallData.sol

```diff
@@ -0,0 +1,67 @@
+/*
+
+  Copyright 2020 ZeroEx Intl.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+*/
+
+pragma solidity ^0.6.5;
+pragma experimental ABIEncoderV2;
+
+import "@0x/contracts-utils/contracts/src/v06/LibBytesV06.sol";
+
+
+/// @dev Library for working with signed calldata.
+library LibSignedCallData {
+    using LibBytesV06 for bytes;
+
+    // bytes4(keccak256('SignedCallDataSignature(bytes)'))
+    bytes4 constant private SIGNATURE_SELECTOR = 0xf86d1d92;
+
+    /// @dev Try to parse potentially signed calldata into its hash and signature
+    ///      components. Signed calldata has signature data appended to it.
+    /// @param callData the raw call data.
+    /// @return callDataHash The hash of the signed callData, or 0x0 if no signature
+    ///         is present.
+    /// @return signature The signature bytes, if present.
+    function parseCallData(bytes memory callData)
+        internal
+        pure
+        returns (bytes32 callDataHash, bytes memory signature)
+    {
+        // Signed calldata has a 70 byte signature appended as:
+        // ```
+        //   abi.encodePacked(
+        //     callData,
+        //     bytes4(keccak256('SignedCallDataSignature(bytes)')),
+        //     signature // 66 bytes
+        //   );
+        // ```
+
+        if (
+            // Signed callData has to be at least 70 bytes long.
```

Comment:
I'm not sure whether optimized Solidity is always guaranteed to always execute `callData.length < 70` and only then `SIGNATURE_SELECTOR != callData.readBytes4(callData.length - 70)`. 🤔

Reply:
From the docs:

> The operators || and && apply the common short-circuiting rules. This means that in the expression f(x) || g(y), if f(x) evaluates to true, g(y) will not be evaluated even if it may have side-effects.

Reply:
Awesome, thanks for looking up!